### ISSUE TITLE
Malli.fn :: Invalid input in query-hash and select-keys-for-hashing

### ISSUE THREAD

**Issue by Philippe-M** [metabase/metabase#37345]
### Describe the bug

Hello, 

In log i've this error :

```
WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name metabase.query-processor.util/query-hash}
WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name metabase.query-processor.util/select-keys-for-hashing}
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "fr",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:121.0) Gecko/20100101 Firefox/121.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.15+10-post-Debian-1deb11u1",
    "java.vendor": "Debian",
    "java.vendor.url": "https://tracker.debian.org/openjdk-11",
    "java.version": "11.0.15",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.15+10-post-Debian-1deb11u1",
    "os.name": "Linux",
    "os.version": "5.10.0-14-amd64",
    "user.language": "fr",
    "user.timezone": "Europe/Paris"
  },
  "metabase-info": {
    "databases": [
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MariaDB",
        "version": "10.5.15-MariaDB-0+deb11u1"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.7.10"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-01-05",
      "tag": "v0.48.2",
      "hash": "e66c075"
    },
    "settings": {
      "report-timezone": "Europe/Paris"
    }
  }
}
```
```


### Severity

no

### Additional context

_No response_

**Comment by Philippe-M** [comment]
Same error after 0.48.3 update :(

The issue is that after a few days, the Java process for Metabase consumes 100% of the CPU, and no queries are successful. The only solution is to restart Metabase.

**Comment by paoliniluis** [comment]
@Philippe-M can you post here when that happens? is it a dashboard? a question? can you give us the json output of that dashboard?

**Comment by Philippe-M** [comment]
@paoliniluis : This message appears in the logs when I display a specific dashboard. I have tested all the queries on this dashboard, and none of them show errors, so the issue is indeed with this dashboard. It consists of 5 queries and 4 filters.

![image](https://github.com/metabase/metabase/assets/4558101/d5b2f982-47bb-43fa-9986-e9c1e8c00d0c)

But how-to export the dashboard in json file ?

**Comment by dnorth98** [comment]
FWIW, we seem to have the same issue as this.  There's so many users and dashboards though that I'm not sure what triggers this.  We were tracking down some performance issues and so paying a little more attention to logs than normal.

I don't know if this shot from our Cloudwatch logs is helpful here but I'd be happy to try and get any other info around this....

Metabase 0.48.3

![image](https://github.com/metabase/metabase/assets/6616703/7995b474-82b4-4268-9710-6976babd998d)

**Comment by paoliniluis** [comment]
Ok found a way to reproduce:
1) set up Metabase 46.8
2) make an x-ray with the accounts table, save it
3) move to 48, go to the dashboard that was saved, see the logs

```
2024-01-17 01:22:50,230 DEBUG middleware.log :: GET /api/collection/3/items 200 22.4 ms (5 DB calls) App DB connections: 0/15 Jetty threads: 4/50 (10 idle, 0 queued) (117 total active threads) Queries in flight: 0 (0 queued)
2024-01-17 01:22:52,441 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,444 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,446 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,448 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,451 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,453 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,458 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,461 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,464 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,467 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,469 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name query-hash}
2024-01-17 01:22:52,471 WARN malli.fn :: Invalid input - Please report this as an issue on Github: ["invalid type"] {:type :metabase.util.malli.fn/invalid-input, :error {:schema :map, :value nil, :errors ({:path [], :in [], :schema :map, :value nil, :type :malli.core/invalid-type})}, :humanized [invalid type], :schema :map, :value nil, :fn-name select-keys-for-hashing}
2024-01-17 01:22:52,480 DEBUG middleware.log :: GET /api/dashboard/1 200 119.8 ms (55 DB calls) App DB connections: 0/15 Jetty threads: 4/50 (10 idle, 0 queued) (117 total active threads) Queries in flight: 0 (0 queued)

```

**Comment by bshepherdson** [comment]
Okay, I can reproduce this. Further I can reproduce this on master in dev mode, which gives a lot more intel about the underlying issue.

The actual error message is coming from `nil` being passed as the query to `metabase.query-processor.util/query-hash`. That's coming from `metabase.api.dashboard/card->query-hashes`, which is trying to compute the hashes to look up the query cards.

This has actually been "failing" for ages; lots of cards (eg. markdown cards) don't have a `:dataset_query`. When the schema violation was throwing an error (before 48.1?) it was getting eaten by `u/ignore-exceptions` in `card->query-hashes`; now that it's logging a warning instead that has become visible.

I'll tweak the logic to handle missing `:dataset_query` properly and the fix will go in 48.4.
